bug: ORDER BY test OFFSET 10; panics the node if test is not indexed

I have observed on my table when doing 
`SELECT * FROM "main_Customer" ORDER BY test LIMIT 10 OFFSET 10`
if test is not indexed column it returns always 0 rows; no matter how full the table is even with 30 rows (it should return 10).
Then I tried it without Limit
`SELECT * FROM "main_Customer" ORDER BY test OFFSET 10`
and my node panics.

If test is indexed everything works just as expected.

**BUG REPORT**

1. Please supply the header (i.e. the first few lines) of your most recent
   log file **for each node in your cluster**. On most unix-based systems
   running with defaults, this boils down to the output of

```
I170523 22:41:50.275998 1 util/log/clog.go:1011  [config] file created at: 2017/05/23 22:41:50
I170523 22:41:50.275998 1 util/log/clog.go:1011  [config] running on machine: bender
I170523 22:41:50.275998 1 util/log/clog.go:1011  [config] binary: CockroachDB CCL v1.0 (darwin amd64, built 2017/05/10 13:17:56, go1.8.1)
I170523 22:41:50.275998 1 util/log/clog.go:1011  [config] arguments: [cockroach start --insecure --host=localhost]
I170523 22:41:50.275998 1 util/log/clog.go:1011  line format: [IWEF]yymmdd hh:mm:ss.uuuuuu goid file:line msg utf8=✓
I170523 22:41:50.275998 1 cli/start.go:575  CockroachDB CCL v1.0 (darwin amd64, built 2017/05/10 13:17:56, go1.8.1)
I170523 22:41:50.379593 7 cli/start.go:338  starting cockroach node
W170523 22:41:50.382712 7 server/server.go:162  running in insecure mode, this is strongly discouraged. See --insecure.
W170523 22:41:50.386473 7 server/config_unix.go:106  soft open file descriptor limit 10240 is under the recommended limit 15000; this may decrease performance
please see https://www.cockroachlabs.com/docs/recommended-production-settings.html for more details
I170523 22:41:50.386737 7 storage/engine/rocksdb.go:380  opening rocksdb instance at "/Users/tvi/go/src/github.com/tvi/gotham/_examples/gorm/cockroach-data"
I170523 22:41:50.426046 7 server/config.go:421  1 storage engine initialized
I170523 22:41:50.426622 7 server/server.go:678  [n?] sleeping for 458.465224ms to guarantee HLC monotonicity
W170523 22:41:50.906472 97 storage/store.go:1339  [n1,s1,r5/1:/Table/{0-11}] could not gossip system config: [NotLeaseHolderError] r5: replica (n1,s1):1 not lease holder; lease holder unknown
I170523 22:41:50.906744 7 storage/store.go:1252  [n1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I170523 22:41:50.906832 7 server/node.go:467  [n1] initialized store [n1,s1]: {Capacity:249779191808 Available:128712523776 RangeCount:70 LeaseCount:0}
I170523 22:41:50.906879 7 server/node.go:351  [n1] node ID 1 initialized
I170523 22:41:50.906996 7 gossip/gossip.go:297  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"localhost:26257" > attrs:<> locality:<> 
I170523 22:41:50.907223 7 storage/stores.go:296  [n1] read 0 node addresses from persistent storage
I170523 22:41:50.908372 7 server/node.go:608  [n1] connecting to gossip network to verify cluster ID...
I170523 22:41:50.908401 7 server/node.go:633  [n1] node connected via gossip and verified as part of cluster "da78a4de-c3ab-4435-9b2c-8cab781af824"
I170523 22:41:50.908454 7 server/node.go:405  [n1] node=1: started with [[]=/Users/tvi/go/src/github.com/tvi/gotham/_examples/gorm/cockroach-data] engine(s) and attributes []
I170523 22:41:50.908638 7 sql/executor.go:338  [n1] creating distSQLPlanner with address {tcp localhost:26257}
I170523 22:41:50.914489 7 server/server.go:741  [n1] starting http server at localhost:8080
I170523 22:41:50.914509 7 server/server.go:742  [n1] starting grpc/postgres server at localhost:26257
I170523 22:41:50.914523 7 server/server.go:743  [n1] advertising CockroachDB node at localhost:26257
I170523 22:41:50.915103 7 server/server.go:853  [n1] done ensuring all necessary migrations have run
I170523 22:41:50.915123 7 server/server.go:855  [n1] serving sql connections
I170523 22:41:50.915229 7 cli/start.go:409  node startup completed:
CockroachDB node starting at 2017-05-23 22:41:50.915147736 +0200 CEST
build:      CCL v1.0 @ 2017/05/10 13:17:56 (go1.8.1)
admin:      http://localhost:8080
sql:        postgresql://root@localhost:26257?sslmode=disable
logs:       /Users/tvi/go/src/github.com/tvi/gotham/_examples/gorm/cockroach-data/logs
store[0]:   path=/Users/tvi/go/src/github.com/tvi/gotham/_examples/gorm/cockroach-data
status:     restarted pre-existing node
clusterID:  da78a4de-c3ab-4435-9b2c-8cab781af824
nodeID:     1
panic: row index 0 out of range

goroutine 478 [running]:
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*RowContainer).At(0xc42098a000, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/row_container.go:268 +0x205
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*rowContainer).MaybeReplaceMax(0xc42098a000, 0xc4216459e8, 0x6, 0x6, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/row_container.go:111 +0x4b
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*sortTopKStrategy).Execute(0xc42098a000, 0x6b9b8e8, 0xc421098330, 0xc421242000, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/sorterstrategy.go:149 +0x10d
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*sorter).Run(0xc421242000, 0x6b9b8e8, 0xc421098330, 0xc420543dd0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/sorter.go:107 +0x2d4
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Start
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:321 +0x337

```

   When log files are not available, supply the output of `cockroach version`
   and all flags/environment variables passed to `cockroach start` instead.

2. Please describe the issue you observed:

- What did you do?
```
SELECT * FROM "main_Customer" ORDER BY ID OFFSET 10;
connection lost; opening new connection and resetting session parameters...
+------------------------------------------+-----------------+-------------+--------------+------+------------------------------------------+
|                    id                    |      email      |    name     | phone_number | test |                test_test                 |
+------------------------------------------+-----------------+-------------+--------------+------+------------------------------------------+
| 6053d3343000fcc545d728765af98635b13b100e | email@email.com | FName LName |            0 | Test | 6053d3343000fcc545d728765af98635b13b100e |

SELECT * FROM "main_Customer" ORDER BY test OFFSET 10;
driver: bad connection
connection lost; opening new connection and resetting session parameters...
```

- What did you expect to see?

- What did you see instead?
Panic - node dies.

Table looks like this

```
CREATE TABLE "main_Customer" (
	id STRING NOT NULL,
	email STRING NULL,
	name STRING NULL,
	phone_number STRING NULL,
	test STRING NULL,
	test_test STRING NULL,
	CONSTRAINT "primary" PRIMARY KEY (id ASC),
	FAMILY "primary" (id, email, name, phone_number, test, test_test)
)
```

Thanks for the report and sorry you're experiencing this difficulty.

I've confirmed your report:

```
root@:26257/test> create table t (a int);
CREATE TABLE
root@:26257/test> insert into t values(1),(2);
INSERT 2
root@:26257/test> SELECT a FROM t ORDER BY a LIMIT 1 offset 1;
+---+
| a |
+---+
+---+
(0 rows)
root@:26257/test> SELECT a FROM t ORDER BY a offset 1;
+---+
| a |
+---+
+---+
(0 rows)
```

The last command produces a panic:
```
panic: row index 0 out of range

goroutine 644 [running]:
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*RowContainer).At(0xc420e12000, 0x0, 0x0, 0x0, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/row_container.go:268 +0x205
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*rowContainer).MaybeReplaceMax(0xc420e12000, 0xc420f799e8, 0x1, 0x1, 0x0, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/row_container.go:111 +0x4b
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*sortTopKStrategy).Execute(0xc420e12000, 0x92235b8, 0xc420c9fc80, 0xc420cee000, 0x0, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/sorterstrategy.go:149 +0x10d
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*sorter).Run(0xc420cee000, 0x92235b8, 0xc420c9fc80, 0xc420563dd0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/sorter.go:107 +0x2d4
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Start
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:321 +0x337
```

Assigning to @RaduBerinde for DistSQL triage.

--- a/pkg/sql/distsqlplan/physical_plan.go
+++ b/pkg/sql/distsqlplan/physical_plan.go
@@ -594,6 +594,8 @@ func (p *PhysicalPlan) AddLimit(count int64, offset int64, node roachpb.NodeID) 
 
 	if len(p.ResultRouters) == 1 {
 		// We only have one processor producing results. Just update its PostProcessSpec.
+		// SELECT FROM (SELECT OFFSET 10 LIMIT 1000) OFFSET 5 LIMIT 20 becomes
+		// SELECT OFFSET 10+5 LIMIT min(1000, 20).
 		post := p.GetLastStagePost()
 		if offset != 0 {
 			if post.Limit > 0 && post.Limit <= uint64(offset) {
@@ -605,7 +607,6 @@ func (p *PhysicalPlan) AddLimit(count int64, offset int64, node roachpb.NodeID) 
 				*p = emptyPlan(p.ResultTypes, node)
 				return nil
 			}
-			post.Limit -= uint64(offset)
 			post.Offset += uint64(offset)
 		}
 		if count != math.MaxInt64 && (post.Limit == 0 || post.Limit > uint64(count)) {

--- a/pkg/sql/distsqlrun/processors.pb.go
+++ b/pkg/sql/distsqlrun/processors.pb.go
@@ -270,7 +270,8 @@ type PostProcessSpec struct {
 	RenderExprs []Expression `protobuf:"bytes,4,rep,name=render_exprs,json=renderExprs" json:"render_exprs"`
 	// If nonzero, the first <offset> rows will be suppressed.
 	Offset uint64 `protobuf:"varint,5,opt,name=offset" json:"offset"`
-	// If nonzero, the processor will stop after emitting this many rows.
+	// If nonzero, the processor will stop after emitting this many rows. The rows
+	// suppressed by <offset>, if any, do not count towards this limit.
 	Limit uint64 `protobuf:"varint,6,opt,name=limit" json:"limit"`
 }
 

--- a/pkg/sql/distsqlrun/processors.proto
+++ b/pkg/sql/distsqlrun/processors.proto
@@ -101,7 +101,8 @@ message PostProcessSpec {
   // If nonzero, the first <offset> rows will be suppressed.
   optional uint64 offset = 5 [(gogoproto.nullable) = false];
 
-  // If nonzero, the processor will stop after emitting this many rows.
+  // If nonzero, the processor will stop after emitting this many rows. The rows
+  // suppressed by <offset>, if any, do not count towards this limit.
   optional uint64 limit = 6 [(gogoproto.nullable) = false];
 }
 

--- a/pkg/sql/distsqlrun/sorter.go
+++ b/pkg/sql/distsqlrun/sorter.go
@@ -39,7 +39,10 @@ type sorter struct {
 	out      procOutputHelper
 	ordering sqlbase.ColumnOrdering
 	matchLen uint32
-	limit    int64
+	// limit is the maximum number of rows that the sorter will push to the
+	// procOutputHelper. 0 if the sorter should sort and push all the rows from
+	// the input.
+	limit int64
 }
 
 var _ processor = &sorter{}
@@ -47,13 +50,19 @@ var _ processor = &sorter{}
 func newSorter(
 	flowCtx *FlowCtx, spec *SorterSpec, input RowSource, post *PostProcessSpec, output RowReceiver,
 ) (*sorter, error) {
+	limit := int64(0)
+	if post.Limit != 0 {
+		// The sorter needs to produce Offset + Limit rows. The procOutputHelper
+		// will discard the first Offset ones.
+		limit = int64(post.Limit) + int64(post.Offset)
+	}
 	s := &sorter{
 		flowCtx:  flowCtx,
 		input:    MakeNoMetadataRowSource(input, output),
 		rawInput: input,
 		ordering: convertToColumnOrdering(spec.OutputOrdering),
 		matchLen: spec.OrderingMatchLen,
-		limit:    int64(post.Limit),
+		limit:    limit,
 	}
 	if err := s.out.init(post, input.Types(), &flowCtx.evalCtx, output); err != nil {
 		return nil, err

--- a/pkg/sql/distsqlrun/sorter.go
+++ b/pkg/sql/distsqlrun/sorter.go
@@ -39,10 +39,10 @@ type sorter struct {
 	out      procOutputHelper
 	ordering sqlbase.ColumnOrdering
 	matchLen uint32
-	// limit is the maximum number of rows that the sorter will push to the
+	// count is the maximum number of rows that the sorter will push to the
 	// procOutputHelper. 0 if the sorter should sort and push all the rows from
 	// the input.
-	limit int64
+	count int64
 }
 
 var _ processor = &sorter{}
@@ -50,11 +50,11 @@ var _ processor = &sorter{}
 func newSorter(
 	flowCtx *FlowCtx, spec *SorterSpec, input RowSource, post *PostProcessSpec, output RowReceiver,
 ) (*sorter, error) {
-	limit := int64(0)
+	count := int64(0)
 	if post.Limit != 0 {
 		// The sorter needs to produce Offset + Limit rows. The procOutputHelper
 		// will discard the first Offset ones.
-		limit = int64(post.Limit) + int64(post.Offset)
+		count = int64(post.Limit) + int64(post.Offset)
 	}
 	s := &sorter{
 		flowCtx:  flowCtx,
@@ -62,7 +62,7 @@ func newSorter(
 		rawInput: input,
 		ordering: convertToColumnOrdering(spec.OutputOrdering),
 		matchLen: spec.OrderingMatchLen,
-		limit:    limit,
+		count:    count,
 	}
 	if err := s.out.init(post, input.Types(), &flowCtx.evalCtx, output); err != nil {
 		return nil, err
@@ -89,7 +89,7 @@ func (s *sorter) Run(ctx context.Context, wg *sync.WaitGroup) {
 	// Construct the optimal sorterStrategy.
 	var ss sorterStrategy
 	if s.matchLen == 0 {
-		if s.limit == 0 {
+		if s.count == 0 {
 			// No specified ordering match length and unspecified limit; no
 			// optimizations are possible so we simply load all rows into memory and
 			// sort all values in-place. It has a worst-case time complexity of
@@ -100,7 +100,7 @@ func (s *sorter) Run(ctx context.Context, wg *sync.WaitGroup) {
 			// our sort procedure by maintaining a max-heap populated with only the
 			// smallest k rows seen. It has a worst-case time complexity of
 			// O(n*log(k)) and a worst-case space complexity of O(k).
-			ss = newSortTopKStrategy(sv, s.limit)
+			ss = newSortTopKStrategy(sv, s.count)
 		}
 	} else {
 		// Ordering match length is specified. We will be able to use existing

--- a/pkg/sql/testdata/logic_test/select
+++ b/pkg/sql/testdata/logic_test/select
@@ -243,6 +243,16 @@ SELECT * FROM xyzw ORDER BY x LIMIT 1 OFFSET 1
 ----
 4 5 6 7
 
+query IIII
+SELECT * FROM xyzw ORDER BY y OFFSET 1
+----
+4 5 6 7
+
+query IIII
+SELECT * FROM xyzw ORDER BY y OFFSET 1 LIMIT 1
+----
+4 5 6 7
+
 # Multiplying by zero so the result is deterministic.
 query IIII
 SELECT * FROM xyzw LIMIT (RANDOM() * 0.0)::int OFFSET (RANDOM() * 0.0)::int

